### PR TITLE
feat(sync): allow to force a full sync anytime

### PR DESCRIPTION
**Pull request by bodinsamuel**
## Context

Fixes NAN-305

Customers want to trigger a full resync without having to delete a sync / connection. We assume they want to get a fresh data set and they might not have `track_deletes` activated. 

## Proposal

There were no proper mechanism in place to achieve that, so I had to modify a lot of logic here and there to make it work.
- `GET /sync/trigger { full_resync: true }`
- Cancel running schedule
- Delete all the records associated with the SyncId
- Clear the lastSyncDate so the scripts can start from scratch
- Trigger an initial Sync

This operation is destructive and if the sync fails the customers will have no data. But it's the intended behavior and not different from what they were doing manually. We can improve this behavior later.

## Describe your changes

- Orchestrator `runSyncCommand` has a different execution path when `fullResync` is passed.
We use the initial sync as a matter of clarity and logging. We don't want to have "Continuous" when it's in fact a "from scratch sync". I also didn't wanted to have a new "workflow" because of concurrency issues; initial sync concurrency is already handled so it limits the logic required to make this works.

- Separate Initial Sync and Init activities into two distinct Activity.  
Currently we were passing `activitLogId` to temporal to have a continuous log. But we ended up storing this `activityLogId` for ever and could lead to mistakes (e.g: writing to the wrong activity log).
In this case, we were logging to the first very Activity which was incorrect. I had to improvise a bit here so let me know what you think.

Screenshot 2024-02-19 at 11 55 32

- Add missing `SYNC_CANCEL` in analytics
Unrelated but I saw it missing

## How to test this?

Use this integrations

```ts
import type { NangoSync } from './models';

export default async function fetchData(nango: NangoSync): Promise<void> {
    if (nango.lastSyncDate) {
        await nango.log('With sync date');
        await nango.batchSave(
            [
                { id: 1, created_at: new Date() },
                { id: 2, created_at: new Date() }
            ],
            'Simple'
        );
    } else {
        await nango.log('No sync date');
        await nango.batchSave([{ id: 1, created_at: new Date() }], 'Simple');
    }

    await nango.log('✅ end');
}
```

- Create this integrations
- It will create 1 record at the first execution, then 2 at each subsequent executions.
- Call /sync/trigger with `full_resync: false` multiple times
- Check DB `_nango_sync_date_records`: you should have 2 records
- Call /sync/trigger with `full_resync: true`
- Check DB `_nango_sync_date_records`: you should have 1 record

### CODE REVIEW

File: packages/jobs/lib/activities.ts
Comment (bodinsamuel) [inline packages/jobs/lib/activities.ts:420]: In this case (sync cancel) we were passing the old activity log which was not creating an issue per say because we were not logging anything here but highlight the issue of this floating `activityLogId`

File: packages/shared/lib/services/sync/sync.service.ts
Comment (bodinsamuel) [inline packages/shared/lib/services/sync/sync.service.ts:128]: I had forgot to remove this unnecessary logic now that we only pass Date in a controlled way

File: packages/shared/lib/services/sync/orchestrator.service.ts
Comment (TBonnin) [inline packages/shared/lib/services/sync/orchestrator.service.ts]: would it make sense to have 2 commands `RUN` and `RUN_FULL`?

**Review by TBonnin — COMMENTED** [review]
lgtm but there is some part like the orchestrator I am not super familiar with. It would be good if @khaliqgant also review this PR

**Comment by khaliqgant** [comment]
> lgtm but there is some part like the orchestrator I am not super familiar with. It would be good if @khaliqgant also review this PR

Next on my list for today

File: packages/shared/lib/services/sync/orchestrator.service.ts
Comment (khaliqgant) [inline packages/shared/lib/services/sync/orchestrator.service.ts:196]: The activity log should update to reflect that is was a "full resync" performed

image

File: packages/shared/lib/services/sync/orchestrator.service.ts
Comment (khaliqgant) [inline packages/shared/lib/services/sync/orchestrator.service.ts]: `failed_to_get_sync_client`

File: packages/shared/lib/services/sync/orchestrator.service.ts
Comment (khaliqgant) [inline packages/shared/lib/services/sync/orchestrator.service.ts]: Probably out of scope but I don't like how this logic is repeated (probably my fault 🤦🏿‍♂️ from earlier implementation)

File: packages/shared/lib/clients/sync.client.ts
Comment (khaliqgant) [inline packages/shared/lib/clients/sync.client.ts]: I realize I originally wrote this messaging but I think it can be improved. It is kind of an internal message that gives the schedule id and the task queue -- both of which the user doesn't care about. I also wonder if now it is necessary to have both an init log message and the sync kick off activity log message if the sync is `auto_start = true`. Did you discuss this with @bastienbeurier at all?

File: packages/shared/lib/clients/sync.client.ts
Comment (khaliqgant) [inline packages/shared/lib/clients/sync.client.ts]: Should we do something here if this call fails?

File: packages/shared/lib/clients/sync.client.ts
Comment (khaliqgant) [inline packages/shared/lib/clients/sync.client.ts:751]: Is the non null assertion needed here if you check for `syncJobId` earlier?

File: packages/shared/lib/services/sync/orchestrator.service.ts
Comment (khaliqgant) [reply packages/shared/lib/services/sync/orchestrator.service.ts]: Agree, the `RUN_FULL` command should be used here: https://github.com/NangoHQ/nango/blob/f718b2ce8edd70de998affc4a85a10475e2b0cee/packages/shared/lib/models/Sync.ts#L258

File: packages/shared/lib/clients/sync.client.ts
Comment (khaliqgant) [reply packages/shared/lib/clients/sync.client.ts]: But I guess because of how you restructured the activityLogId to not live in temporal which I agree is an improvement you need some kind of init activity log to both create and conclude -- so I'm fine with this besides the messaging needing some updates.

File: packages/shared/lib/services/sync/orchestrator.service.ts
Comment (khaliqgant) [inline packages/shared/lib/services/sync/orchestrator.service.ts]: Seems like this logic can be abstracted into another function within the orchestrator service

File: packages/jobs/lib/activities.ts
Comment (khaliqgant) [inline packages/jobs/lib/activities.ts]: Should it be full or just go back to being an `INITIAL`? Locally I'm also seeing inconsistent state between doing a full resync and sometimes it being a FULL and sometimes being an INITIAL. I feel like it is easier and less to manage to just have it go back to `INITIAL`

**Review by khaliqgant — CHANGES_REQUESTED** [review]
Overall looks good and mostly works as intended! Feedback inline

File: packages/shared/lib/services/sync/orchestrator.service.ts
Comment (bodinsamuel) [reply packages/shared/lib/services/sync/orchestrator.service.ts]: The method takes an optional `connectionId` it creates two paths in this method

File: packages/shared/lib/clients/sync.client.ts
Comment (bodinsamuel) [reply packages/shared/lib/clients/sync.client.ts:751]: The type `Job` is incorrect it marks id as optional, I guess because the same type is used for creation and for response. 
This is something to fix in a types improvement initiative I think

File: packages/shared/lib/services/sync/orchestrator.service.ts
Comment (bodinsamuel) [reply packages/shared/lib/services/sync/orchestrator.service.ts]: There is already a RUN_FULL command that seems not implemented (or at least I didn't get it) so I was not comfortable using it. Maybe @khaliqgant knows better

edit: I didn't refreshed my page before commenting

File: packages/jobs/lib/activities.ts
Comment (bodinsamuel) [reply packages/jobs/lib/activities.ts]: yep that's a leftover from my early experimentation, good catch

File: packages/shared/lib/clients/sync.client.ts
Comment (bodinsamuel) [reply packages/shared/lib/clients/sync.client.ts]: we were not doing anything else before but we can throw yes

File: packages/shared/lib/services/sync/orchestrator.service.ts
Comment (bodinsamuel) [reply packages/shared/lib/services/sync/orchestrator.service.ts:196]: is it okay I had this in the sentence instead of changing the command? (e.g: `Sync was updated with command: "trigger sync" (full)`)
because creating a new command will have a much bigger impact

File: packages/shared/lib/clients/sync.client.ts
Comment (bodinsamuel) [reply packages/shared/lib/clients/sync.client.ts]: I didn't discuss it yet, I was waiting for a technical approval first to be sure since I have been juggling with 5 different way of doing this. 
I don't have strong opinion on this, but I think transparency is always helpful, especially regarding a free form input like scheduling, so I could just rewrite this msg to a simple `Scheduled to run ${syncData.runs}`, wdyt?

File: packages/shared/lib/services/sync/orchestrator.service.ts
Comment (bodinsamuel) [reply packages/shared/lib/services/sync/orchestrator.service.ts:196]: nvm let's use RUN_FULL then

File: packages/jobs/lib/activities.ts
Comment (bodinsamuel) [reply packages/jobs/lib/activities.ts]: actually after changing to use FULL_RUN, might make more sense to stick with FULL.
INITIAL does not make a lot of sense when we think about it

File: packages/shared/lib/services/sync/orchestrator.service.ts
Comment (bodinsamuel) [reply packages/shared/lib/services/sync/orchestrator.service.ts:196]: Changed ed3472d

File: packages/shared/lib/services/sync/orchestrator.service.ts
Comment (bodinsamuel) [reply packages/shared/lib/services/sync/orchestrator.service.ts]: Changed ed3472d

File: packages/shared/lib/clients/sync.client.ts
Comment (bodinsamuel) [reply packages/shared/lib/clients/sync.client.ts]: Changed ed3472d

File: packages/shared/lib/clients/sync.client.ts
Comment (bodinsamuel) [reply packages/shared/lib/clients/sync.client.ts]: Changed ed3472d

File: packages/shared/lib/services/sync/orchestrator.service.ts
Comment (bodinsamuel) [reply packages/shared/lib/services/sync/orchestrator.service.ts]: Changed ed3472d

File: packages/jobs/lib/activities.ts
Comment (bodinsamuel) [reply packages/jobs/lib/activities.ts]: Changed ed3472d to INITIAL (so my prev comment is still relevant)

**Comment by bodinsamuel** [comment]
Updated with all your comments.
Screenshot 2024-02-20 at 16 39 12

I believe INITIAL run should be entirely dropped in favor of INCREMENTAL or FULL but that needs a little bit more tweaks that might be done later

File: packages/shared/lib/clients/sync.client.ts
Comment (khaliqgant) [reply packages/shared/lib/clients/sync.client.ts:751]: Gotcha, agreed

**Comment by khaliqgant** [comment]
Thanks for integrating the feedback. From a code perspective looks good. I'll give this another check locally tomorrow first thing!

**Review by khaliqgant — CHANGES_REQUESTED** [review]
Seeing a regular trigger sync icon and text missing

image

**Comment by khaliqgant** [comment]
Edge case:
1. Handled: User triggers a sync, then triggers a full resync, the "regular sync" is cancelldd
2. User triggers a full sync, then triggers a "regular sync" -- what happens?

**Comment by bodinsamuel** [comment]
> Edge case:

> 2. User triggers a full sync, then triggers a "regular sync" -- what happens?

We already have the check `activities.ts:80` => `isInitialSyncStillRunning()` 
That's why I'm using INITIAL.

**Review by khaliqgant — APPROVED** [review]

**Pull request merged**
